Release package hold

jcbantuelle opened the pull request:
Thank you for this cookbook, it's been very helpful as a starting point for some server configuration I've been doing.

I just attempted to leverage it to upgrade my existing Dokku install though, and ran into the following error:

```STDERR: E: Held packages were changed and -y was used without --allow-change-held-packages.```

Looking through the codebase, I noticed that after the packages are installed a hold is placed on them. My understanding is that we would need to release that hold prior to attempting any further modification of the packages, which is why I've put together this PR to do so.

While making those changes, I also updated the default version numbers since it's progressed quite a while since the last update, and added test support for Ubuntu 18.04.

I hope this helps and is a welcome addition, but please let me know if you have any reservations or feedback at all and I'll be happy to address.

jcbantuelle committed:
Release Package Hold Before Updating

nickcharlton commented:
Ah, nice idea!

Would you be able to break out the Ubuntu 18.04 and Dokku package version changes? If we open these in two other PRs we can merge those first as they're not related to this change.

It was a while ago I used this, but I think my intention was that you'd put a hold on the package so that you didn't accidentally upgrade (and break) something when you didn't mean to. What does this change in that regard? Perhaps explicitly upgrading the cookbook is enough?

jcbantuelle commented:
Thanks for the quick feedback! I've updated the source branch accordingly to split things out.

Putting a hold on the package makes total sense, and I believe serves the purpose of preventing any automated package updates from introducing a breaking Dokku upgrade.

The issue I encountered was that I needed to upgrade to a newer version of Dokku so I updated my attributes file, but then the chef scripts failed when running the `dokku::install` recipe because of the hold.

My understanding is that updating that attributes file should allow the recipe to properly update the packages when converging, but that the hold prevents any future changes. I figured that if the install recipe removes the hold immediately before attempting to install the dokku package, then reapplies it afterwards, that would enforce the only upgrade path being through an explicit update and meet the need.

nickcharlton commented:
Somewhere the stars aligned where you opened the PR during a block of time I'd reserved for reviewing open source PRs. I'm not usually this fast!

Yeah, that's reasonable! I'm going to go ahead and merge this.

jcbantuelle commented:
That's incredibly fortunate, thank you again!

Do you think you'll have a chance to package this into a new release as well? No worries at if not, I'm just trying to figure out if I should temporarily load from my own version.

nickcharlton commented:
I'll take a look tomorrow; it shouldn't be too hard, even though it's been a while.

I did notice that CI was erroring; if you'd have the time to look at that, that'd be amazing?

jcbantuelle commented:
Hey, I just wanted to check in and see if a new release was still part of your short-term plan. I can bypass for now if not, but figured it'd be more straightforward to pull directly from chef supermarket if possible. Thanks again for the quick review and feedback last week!

nickcharlton commented:
Yep, it is! I don't have a Chef environment working any more, so I need to start with sorting that out first. It'll either be today or tomorrow when it gets released. I'll update here when I have!

nickcharlton commented:
…that was particularly difficult and time consuming but here we go: https://supermarket.chef.io/cookbooks/dokku/versions/2.3.0

jcbantuelle commented:
I'm sorry that was such an ordeal :( I greatly appreciate you persevering however, thank you for all your work!
